addons/sale/models/product_template.py

```diff
@@ -64,19 +66,27 @@ def _compute_sales_count(self):
     def _check_sale_product_company(self):
         """Ensure the product is not being restricted to a single company while
         having been sold in another one in the past, as this could cause issues."""
-        target_company = self.company_id
-        if target_company:  # don't prevent writing `False`, should always work
-            product_data = self.env['product.product'].sudo().with_context(active_test=False).search_read([('product_tmpl_id', 'in', self.ids)], fields=['id'])
-            product_ids = list(map(lambda p: p['id'], product_data))
-            so_lines = self.env['sale.order.line'].sudo().search_read([('product_id', 'in', product_ids), ('company_id', '!=', target_company.id)], fields=['id', 'product_id'])
-            used_products = list(map(lambda sol: sol['product_id'][1], so_lines))
+        products_by_compagny = defaultdict(lambda: self.env['product.template'])
+        for product in self:
+            if not product.product_variant_ids or not product.company_id:
+                # No need to check if the product has just being created (`product_variant_ids` is
+                # still empty) or if we're writing `False` on its company (should always work.)
+                continue
+            products_by_compagny[product.company_id] |= product
+
+        for target_company, products in products_by_compagny.items():
+            subquery_products = self.env['product.product'].sudo().with_context(active_test=False)._search([('product_tmpl_id', 'in', products.ids)])
+            so_lines = self.env['sale.order.line'].sudo().search_read(
+                [('product_id', 'in', subquery_products), '!', ('company_id', 'child_of', target_company.id)],
```

Comment:
```suggestion
                [('product_id', 'in', subquery_products), ('company_id', '!=', target_company.id)],
```
`child_of` only applies in 17.0+, the domain shouldn't be changed here.

Reply:
Thanks for your review @Feyensv, changed the commit accordingly